src/engine/sparqlExpressions/LangExpression.h

```diff
@@ -1,39 +1,63 @@
-//  Copyright 2022, University of Freiburg,
+//  Copyright 2022-2024 University of Freiburg,
 //                  Chair of Algorithms and Data Structures.
-//  Author: Johannes Kalmbach <kalmbach@cs.uni-freiburg.de>
+//  Author: Johannes Kalmbach <kalmbach@cs.uni-freiburg.de> (2022)
+//  Author: Hannes Baumann <baumannh@informatik.uni-freiburg.de> (2024)
 
-// Test file for the LangExpression class: LanguageExpressionsTests.h
+// Unit tests for the functionality from this file can be
+// found in LanguageExpressionsTest.h
 
 #pragma once
 
-#include "engine/sparqlExpressions/SparqlExpression.h"
+#include "./engine/sparqlExpressions/LiteralExpression.h"
+#include "./engine/sparqlExpressions/NaryExpressionImpl.h"
+#include "./engine/sparqlExpressions/SparqlExpressionTypes.h"
 
 namespace sparqlExpression {
-
-class LangExpression : public SparqlExpression {
- private:
-  // The stored variable.
-  Variable variable_;
-  SparqlExpression::Ptr child_;
-
+namespace detail::langImpl {
+
+using Lit = ad_utility::triple_component::Literal;
+using OptValue = std::optional<std::string>;
+
+//______________________________________________________________________________
+// `LangExpressionImpl` extends the `NaryExpression` class with the methods
+// `containsLangExpression()` and `variable()`, which are required for the
+// usage within the `Filter()`.
+template <typename NaryOperation>
+requires(isOperation<NaryOperation>)
+class LangExpressionImpl : public NaryExpression<NaryOperation> {
  public:
-  // Construct from a child expression. The child must be a single variable,
-  // otherwise an exception will be thrown.
-  LangExpression(SparqlExpression::Ptr child);
-
-  const Variable& variable() const { return variable_; }
+  using NaryExpression<NaryOperation>::NaryExpression;
 
   bool containsLangExpression() const override { return true; }
 
-  ExpressionResult evaluate(EvaluationContext* context) const override;
-
-  std::string getCacheKey(const VariableToColumnMap&) const override;
+  const Variable& variable() const {
+    std::optional<SparqlExpression*> optChild = this->getNthChild(0);
+    AD_CORRECTNESS_CHECK(optChild.has_value());
+    if (auto stringPtr =
+            dynamic_cast<const VariableExpression*>(optChild.value())) {
+      return stringPtr->value();
+    } else {
+      throw std::runtime_error{
+          "use LANG() with ?var as an argument within a Filter-expression"};
```

Comment:
Can this error occur with a query (then it should be more understandable), 
Or can it only occur with a programming error (then it would be an `AD_CONTRACT_CHECK(stringPtr != nullptr, "Explanation for the failure")`.

Reply:
Yes it can be triggered with a query, the error message is more explicit now.